File: ssh_sessions/organization.py
"""Grouping, searching and atomic bulk edits for the machine catalog."""
from dataclasses import replace
import shlex

from .catalog import CatalogError, group_path, parse_tags


def within(group, parent):
    return group.casefold() == parent.casefold() or bool(parent) and group.casefold().startswith(parent.casefold() + '/')


def groups(machines):
    counts = {}
    canonical = {}
    for machine in machines:
        parts = machine.group.split('/') if machine.group else []
        for depth in range(1, len(parts) + 1):
            path = '/'.join(parts[:depth])
            key = canonical.setdefault(path.casefold(), path)
            counts[key] = counts.get(key, 0) + 1
    return dict(sorted(counts.items(), key=lambda item: item[0].casefold()))


def matches(machine, query):
    haystack = ' '.join([machine.name, machine.user, machine.group, *machine.tags,
                        *(r.host for r in machine.routes), *(r.name for r in machine.routes),
                        *(r.ssh_alias or '' for r in machine.routes)]).casefold()
    try:
        terms = shlex.split(query)
    except ValueError:
        terms = query.split()
    for term in terms:
        if term.casefold().startswith('tag:'):
            if term[4:].casefold() not in {tag.casefold() for tag in machine.tags}:
                return False
        elif term.casefold().startswith('group:'):
            if not within(machine.group, term[6:]):
                return False
        elif term.casefold() not in haystack:
            return False
    return True


def filtered(machines, query='', group=None, tag=None):
    return [m for m in machines if (group is None or within(m.group, group)) and matches(m, query)
            and (tag is None or tag.casefold() in {t.casefold() for t in m.tags})]


def edit_many(catalog, machine_ids, expected, *, group=None, add_tags=(), remove_tags=()):
    snapshot = catalog.load()
    ids = set(machine_ids)
    if snapshot.revision != expected:
        raise CatalogError('The catalog changed. Reload before applying this edit.')
    if not ids or ids - {m.id for m in snapshot.machines}:
        raise CatalogError('Select machines from the current catalog.')
    if group is not None:
        group = group_path(group)
        group = next((name for name in groups(snapshot.machines) if name.casefold() == group.casefold()), group)
    added = parse_tags(add_tags)
    removed = {tag.casefold() for tag in parse_tags(remove_tags)}
    if removed & {tag.casefold() for tag in added}:
        raise CatalogError('A tag cannot be added and removed in the same edit.')
    changed = tuple(replace(m, group=m.group if group is None else group,
                            tags=parse_tags([t for t in m.tags if t.casefold() not in removed] + list(added)))
                    if m.id in ids else m for m in snapshot.machines)
    return catalog.save(changed, expected)


def rename_group(catalog, old, new, expected):
    old, new = group_path(old), group_path(new)
    snapshot = catalog.load()
    if not old or not new or not any(within(m.group, old) for m in snapshot.machines):
        raise CatalogError('Choose an existing group and a nonempty new name.')
    if new.casefold().startswith(old.casefold() + '/'):
        raise CatalogError('A group cannot be moved inside itself.')
    existing = groups(snapshot.machines)
    if old.casefold() != new.casefold() and any(g.casefold() == new.casefold() for g in existing):
        raise CatalogError('That group already exists. Use Move to merge machines into it.')
    def renamed(path):
        suffix = '/'.join(path.split('/')[len(old.split('/')):])
        return group_path(new + ('/' + suffix if suffix else ''))
    changed = tuple(replace(m, group=renamed(m.group)) if within(m.group, old) else m
                    for m in snapshot.machines)
    return catalog.save(changed, expected)

File: src/organization.rs
use crate::catalog::{Catalog, Machine, Snapshot, group_path, tags};
use crate::text::casefold;
use anyhow::{Result, ensure};
use std::collections::{BTreeMap, HashSet};

pub fn within(group: &str, parent: &str) -> bool {
    let group = casefold(group);
    let parent = casefold(parent);
    group == parent || !parent.is_empty() && group.starts_with(&(parent + "/"))
}
pub fn groups(machines: &[Machine]) -> BTreeMap<String, usize> {
    let mut result = BTreeMap::new();
    let mut canonical = BTreeMap::new();
    for machine in machines {
        if machine.group.is_empty() {
            continue;
        }
        let parts: Vec<_> = machine.group.split('/').collect();
        for depth in 1..=parts.len() {
            let name = parts[..depth].join("/");
            let key = canonical.entry(casefold(&name)).or_insert(name);
            *result.entry(key.clone()).or_insert(0) += 1;
        }
    }
    result
}
pub fn matches(machine: &Machine, query: &str) -> bool {
    let mut words = vec![
        machine.name.clone(),
        machine.user.clone(),
        machine.group.clone(),
    ];
    words.extend(machine.tags.clone());
    for route in &machine.routes {
        words.extend([
            route.host.clone(),
            route.name.clone(),
            route.ssh_alias.clone().unwrap_or_default(),
        ]);
    }
    let haystack = casefold(&words.join(" "));
    shlex::split(query)
        .unwrap_or_else(|| query.split_whitespace().map(str::to_owned).collect())
        .iter()
        .all(|term| {
            let term = casefold(term);
            if let Some(tag) = term.strip_prefix("tag:") {
                machine.tags.iter().any(|t| casefold(t) == tag)
            } else if let Some(group) = term.strip_prefix("group:") {
                within(&machine.group, group)
            } else {
                haystack.contains(&term)
            }
        })
}
pub fn filtered<'a>(
    machines: &'a [Machine],
    query: &str,
    group: Option<&str>,
    tag: Option<&str>,
) -> Vec<&'a Machine> {
    machines
        .iter()
        .filter(|m| {
            group.is_none_or(|g| within(&m.group, g))
                && matches(m, query)
                && tag.is_none_or(|t| m.tags.iter().any(|s| casefold(s) == casefold(t)))
        })
        .collect()
}
pub fn edit_many(
    catalog: &Catalog,
    ids: &[String],
    expected: &str,
    group: Option<&str>,
    add: &[String],
    remove: &[String],
) -> Result<Snapshot> {
    let mut snapshot = catalog.load()?;
    ensure!(
        snapshot.revision == expected,
        "Catalog changed. Reload before editing."
    );
    ensure!(
        !ids.is_empty()
            && ids
                .iter()
                .all(|id| snapshot.machines.iter().any(|m| &m.id == id)),
        "Select machines from the current catalog."
    );
    let group = group.map(group_path).transpose()?.map(|g| {
        groups(&snapshot.machines)
            .into_keys()
            .find(|old| casefold(old) == casefold(&g))
            .unwrap_or(g)
    });
    let add = tags(add.to_vec())?;
    let remove: HashSet<_> = tags(remove.to_vec())?
        .into_iter()
        .map(|t| casefold(&t))
        .collect();
    ensure!(
        !add.iter().any(|t| remove.contains(&casefold(t))),
        "A tag cannot be added and removed in the same edit."
    );
    for machine in &mut snapshot.machines {
        if ids.contains(&machine.id) {
            if let Some(group) = &group {
                machine.group = group.clone();
            }
            machine.tags.retain(|t| !remove.contains(&casefold(t)));
            machine.tags.extend(add.clone());
            machine.tags = tags(machine.tags.clone())?;
        }
    }
    catalog.save(&snapshot.machines, expected)
}
pub fn rename_group(catalog: &Catalog, old: &str, new: &str, expected: &str) -> Result<Snapshot> {
    let old = group_path(old)?;
    let new = group_path(new)?;
    let mut snapshot = catalog.load()?;
    ensure!(
        !old.is_empty()
            && !new.is_empty()
            && snapshot.machines.iter().any(|m| within(&m.group, &old)),
        "Choose an existing group and a nonempty new name."
    );
    ensure!(
        !casefold(&new).starts_with(&(casefold(&old) + "/")),
        "A group cannot be moved inside itself."
    );
    ensure!(
        casefold(&old) == casefold(&new)
            || !groups(&snapshot.machines)
                .keys()
                .any(|g| casefold(g) == casefold(&new)),
        "That group exists. Use Move to merge machines."
    );
    for machine in &mut snapshot.machines {
        if within(&machine.group, &old) {
            let suffix = machine
                .group
                .split('/')
                .skip(old.split('/').count())
                .collect::<Vec<_>>()
                .join("/");
            machine.group = group_path(&format!(
                "{new}{}{suffix}",
                if suffix.is_empty() { "" } else { "/" }
            ))?;
        }
    }
    catalog.save(&snapshot.machines, expected)
}
